ttyplot.c

```diff
@@ -581,33 +549,33 @@ int main(int argc, char *argv[]) {
         cbreak();
     }
 
-    sigemptyset(&empty_sigset);
-
-    sigemptyset(&block_sigset);
-    sigaddset(&block_sigset, SIGINT);
-    sigaddset(&block_sigset, SIGWINCH);
-    sigprocmask(SIG_BLOCK, &block_sigset, NULL);
+    signal(SIGWINCH, signal_handler);
+    signal(SIGINT, signal_handler);
 
-    signal(SIGWINCH, resize);
-    signal(SIGINT, finish);
+    if (pipe(signals) != 0) {
+        perror("pipe");
+        exit(1);
+    }
 
     while(1) {
         // Block until (a) we receive a signal or (b) stdin can be read without blocking
         // or (c) timeout expires, in order to reduce use of CPU and power while idle
         fd_set read_fds;
         FD_ZERO(&read_fds);
-        int select_nfds = 0;
+        FD_SET(signals[0], &read_fds);
+        int select_nfds = signals[0] + 1;
         if (stdin_is_open) {
             FD_SET(STDIN_FILENO, &read_fds);
-            select_nfds = STDIN_FILENO + 1;
+            if (STDIN_FILENO >= select_nfds)
+                select_nfds = STDIN_FILENO + 1;
         }
         if (tty != -1) {
             FD_SET(tty, &read_fds);
             if (tty >= select_nfds)
                 select_nfds = tty + 1;
         }
-        struct timespec timeout = { .tv_sec = 0, .tv_nsec = 500e6 };  // 500 milliseconds for refreshing the clock
-        const int select_ret = pselect_without_signal_starvation(select_nfds, &read_fds, NULL, NULL, &timeout, &empty_sigset);
+        struct timeval timeout = { .tv_sec = 0, .tv_usec = 500e3 };  // 500 milliseconds for refreshing the clock
+        const int select_ret = select(select_nfds, &read_fds, NULL, NULL, &timeout);
```

Comment:
@edgar-bonet I only notice now that with this falling back to plain `select` rather than `pselect`, this would not fix the race but just move the race elsewhere? Am I missing something?

Also, if we do blocking reads and writes to the pipe, isn't that a problem? Do we need to put it in non-blocking mode and read as many signals as there are rather than just one?

Reply:
I see no race here. Signals are never blocked, so the `write` to the pipe will always happen as soon as a signal arrives. And the signal will be handled on the next loop iteration.

Blocking reads should not block, as per the `select` promise. We may put all the file descriptors in non-blocking mode if we want to be extra defensive. Writes could only block if the pipe gets full, which can only happen if you send several thousand signals per second. You may be able to do that programmatically, but not by resizing the terminal or typing Ctrl-C.

Reply:
> [..] Signals are never blocked [..]

@edgar-bonet while you say it like that, do we need to bring back signal blocking around interaction with ncurses a la pull request #130 then?

> [T]he `write` to the pipe will always happen as soon as a signal arrives. And the signal will be handled on the next loop iteration.
> 
> Blocking reads should not block, as per the `select` promise.

Understood, yes.

> We may put all the file descriptors in non-blocking mode if we want to be extra defensive. Writes could only block if the pipe gets full, which can only happen if you send several thousand signals per second. You may be able to do that programmatically, but not by resizing the terminal or typing Ctrl-C.

I mis-remembered then probably: I thought that writes to a pipe always block until there is a reader :thinking: `man 7 pipe` is with you.

Reply:
> do we need to bring back signal blocking around interaction with ncurses a la pull request #130 then?

We don't. The signal handler does nothing but a `write`, which is async-signal-safe as per POSIX. The signal can really interrupt us any time without disrupting the program.

Reply:
@edgar-bonet isn't the important question whether the ncurses functions will stop their job half way when interrupted by a signal or not? From a quick look the man pages of `erase` and `refresh` say nothing about signals. I think that's what we'd need to know to be sure of no disruption. What do you think?

Reply:
If an ncurses function is interrupted by a signal, it will stop halfway... then resume from where it stopped, and not even be aware that it had been interrupted. It is not very different from what happens to any running program during a hardware interrupt.

Most of these functions are not async-signal-safe, but this does _not_ mean it is unsafe to interrupt them: it means it is unsafe to call them from a signal handler.

That being said, it is conceivable that some ncurses function is buggy and mishandles a syscall failing with `EINTR`. In that case it would make sense to work around that bug by blocking signals around the call. However, I would avoid implementing workarounds for bugs we do not even know they exist.

Reply:
@edgar-bonet okay, let's note though that when we learn about that later, it will have been this move back to `select` from `pselect` that caused it and that — compared to 1.5.2 — now we have `select` without blocking signals while we had `select` *with* blocking signals in 1.5.2. There are multiple ways to see this and so going forward with #143 means that you are taking responsibility for dropping these blocks compared to 1.5.2; I cannot take the responsibility, because I cannot guarantee, that it will not cause new problems. Is that aware and okay with you? I'll do a last round of review then…

Reply:
> we had `select` _with_ blocking signals in 1.5.2.

Let me take a look at 1.5.2 again... I see. That version called ncurses from the `SIGWINCH` signal handler, which is inherently unsafe and can lead, e.g., to memory corruption. This is reported in issue #62.

That version also blocked signals around calls to ncurses. This is a reasonable mitigation strategy: it prevents ncurses from being called on top of an interrupted ncurses function, and should thus prevent ncurses from corrupting its own state. The mitigation is not perfect though, and not guaranteed to always work. For example, it cannot prevent ncurses from calling `malloc` on top of an interrupted `malloc` call.

Now that we have a signal handler that is completely async-signal-safe, that mitigation has lost its purpose.

Reply:
@edgar-bonet I'll take that as "I'm sure enough that a block-do-unblock wrapper is not needed to take full responsibility for not re-adding some". Good.